File: internal/sam/wire.go
package sam

import (
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"io"
	"net"
	"time"
)

// SendPayload пишет в conn: size-header + payload из reader + SHA256.
func SendPayload(conn net.Conn, r io.Reader, size int64) (int64, [32]byte, error) {
	// 1. Write size header (8 bytes big-endian)
	if err := binary.Write(conn, binary.BigEndian, uint64(size)); err != nil {
		return 0, [32]byte{}, fmt.Errorf("failed to write size: %v", err)
	}

	// 2. Write payload and calculate SHA256
	h := sha256.New()
	mw := io.MultiWriter(conn, h)
	written, err := io.CopyN(mw, r, size)
	if err != nil {
		return written, [32]byte{}, fmt.Errorf("failed to write payload: %v", err)
	}

	var sum [32]byte
	copy(sum[:], h.Sum(nil))

	// 3. Write SHA256 (32 bytes)
	if _, err := conn.Write(sum[:]); err != nil {
		return written, sum, fmt.Errorf("failed to write sha256: %v", err)
	}

	return written, sum, nil
}

// ReceivePayload читает из conn size-header, затем payload,
// затем 32 байта SHA256 от sender'а.
func ReceivePayload(conn net.Conn) (received int64, sha256ok bool, firstByteAt time.Time, err error) {
	// 1. Read size header
	var size uint64
	if err := binary.Read(conn, binary.BigEndian, &size); err != nil {
		return 0, false, time.Time{}, fmt.Errorf("failed to read size: %v", err)
	}

	// 2. Read first byte of payload to fix firstByteAt
	h := sha256.New()
	var buf [1]byte
	if _, err := io.ReadFull(conn, buf[:]); err != nil {
		return 0, false, time.Time{}, fmt.Errorf("failed to read first byte: %v", err)
	}
	firstByteAt = time.Now()
	h.Write(buf[:])

	// 3. Read rest of payload
	received = 1
	if size > 1 {
		rest := int64(size - 1)
		nr, err := io.CopyN(h, conn, rest)
		received += nr
		if err != nil {
			return received, false, firstByteAt, fmt.Errorf("failed to read payload: %v", err)
		}
	}

	// 4. Read SHA256 from sender
	var senderSum [32]byte
	if _, err := io.ReadFull(conn, senderSum[:]); err != nil {
		return received, false, firstByteAt, fmt.Errorf("failed to read sha256: %v", err)
	}

	// 5. Verify
	mySum := h.Sum(nil)
	sha256ok = true
	for i := 0; i < 32; i++ {
		if senderSum[i] != mySum[i] {
			sha256ok = false
			break
		}
	}

	return received, sha256ok, firstByteAt, nil
}

File: sam3/src/wire.rs
use sha2::{Digest, Sha256};
use std::{
    error::Error,
    io::{self, Read, Write},
    time::Instant,
};

pub fn send_payload(
    conn: &mut impl Write,
    reader: &mut impl Read,
    size: i64,
) -> Result<(i64, [u8; 32]), Box<dyn Error>> {
    conn.write_all(&(size as u64).to_be_bytes())?;
    let mut h = Sha256::new();
    let mut limited = reader.take(size as u64);
    let written = io::copy(
        &mut limited,
        &mut TeeWriter {
            inner: conn,
            hasher: &mut h,
        },
    )? as i64;
    if written != size {
        return Err(format!("short payload reader: wrote {written}, expected {size}").into());
    }
    let sum: [u8; 32] = h.finalize().into();
    conn.write_all(&sum)?;
    conn.flush()?;
    Ok((written, sum))
}

pub fn receive_payload(conn: &mut impl Read) -> Result<(i64, bool, Instant), Box<dyn Error>> {
    let mut size_buf = [0u8; 8];
    conn.read_exact(&mut size_buf)
        .map_err(|e| format!("failed to read size: {e}"))?;
    let size = u64::from_be_bytes(size_buf);

    let mut h = Sha256::new();
    let mut first = [0u8; 1];
    conn.read_exact(&mut first)
        .map_err(|e| format!("failed to read first byte: {e}"))?;
    let first_byte_at = Instant::now();
    h.update(first);

    let mut received = 1i64;
    if size > 1 {
        let mut limited = conn.take(size - 1);
        let nr = io::copy(&mut limited, &mut h)? as i64;
        received += nr;
        if nr != (size - 1) as i64 {
            return Err(format!("failed to read payload: short read {nr}/{}", size - 1).into());
        }
    }

    let mut sender_sum = [0u8; 32];
    conn.read_exact(&mut sender_sum)
        .map_err(|e| format!("failed to read sha256: {e}"))?;
    let my_sum: [u8; 32] = h.finalize().into();
    Ok((received, sender_sum == my_sum, first_byte_at))
}

struct TeeWriter<'a, W: Write> {
    inner: &'a mut W,
    hasher: &'a mut Sha256,
}

impl<W: Write> Write for TeeWriter<'_, W> {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        let n = self.inner.write(buf)?;
        self.hasher.update(&buf[..n]);
        Ok(n)
    }

    fn flush(&mut self) -> io::Result<()> {
        self.inner.flush()
    }
}

#[cfg(test)]
mod tests {
    use super::{receive_payload, send_payload};
    use crate::payload::{expected_sha256, PayloadReader};
    use std::io::Cursor;

    #[test]
    fn send_and_receive_payload_roundtrip() {
        let mut encoded = Vec::new();
        let mut reader = PayloadReader::new(4096, 42);

        let (written, sent_sum) = send_payload(&mut encoded, &mut reader, 4096).expect("send");
        let (received, sha_ok, _) = receive_payload(&mut Cursor::new(encoded)).expect("receive");

        assert_eq!(written, 4096);
        assert_eq!(received, 4096);
        assert!(sha_ok);
        assert_eq!(sent_sum, expected_sha256(4096, 42));
    }

    #[test]
    fn send_payload_rejects_short_reader() {
        let mut encoded = Vec::new();
        let mut reader = Cursor::new(vec![1, 2, 3]);

        let err = send_payload(&mut encoded, &mut reader, 4).expect_err("short reader");

        assert!(err.to_string().contains("short payload reader"));
    }

    #[test]
    fn receive_payload_rejects_bad_checksum() {
        let mut encoded = Vec::new();
        let mut reader = PayloadReader::new(32, 42);
        send_payload(&mut encoded, &mut reader, 32).expect("send");
        let last = encoded.last_mut().expect("checksum byte");
        *last ^= 0xff;

        let (_, sha_ok, _) = receive_payload(&mut Cursor::new(encoded)).expect("receive");

        assert!(!sha_ok);
    }
}
